### ISSUE TITLE
Bug: TSOA does not output valid type names in Swagger definition

### ISSUE THREAD

**Issue by KennyLindahl** [lukeautry/tsoa#1559]
## Sorting

- **I'm submitting a ...**
  - [x] bug report
  
## Expected Behavior
Should resolve type to acceptable name:

1. Get the ideal name `SearchableField_(typeofshopsSubFields)[ngram]_`
2. Replace unexpected characters to match `^[a-zA-Z0-9\.\-_]+$`

## Current Behavior
```
[23] build/open-api-definition.json:8:3 at #/components/schemas/SearchableField_(typeofshopsSubFields)%5Bngram%5D_

The map key in schemas "SearchableField_(typeofshopsSubFields)%5Bngram%5D_" does not match the regular expression "^[a-zA-Z0-9\.\-_]+$"

 6 |            "requestBodies": {},
 7 |            "responses": {},
 8 |            "schemas": {
 9 |                    "TotalHitsRelation": {
10 |                            "type": "string",

Error was generated by the spec-components-invalid-map-name rule.
```

The types looks like this:
> Type gymnastics - there is pretty much no other way to make this work with TSOA):

```ts
export type ValueOf<T> = T[keyof T];

export type Source = string[] | string;

export const MARKET_IDENTIFIER = '{MARKET}';
export const searchTimeAnalyzer = {
  stemmedSynonyms: `search_analyzer_stemmed_synonyms_${MARKET_IDENTIFIER}`,
  synonyms: `search_analyzer_synonyms_${MARKET_IDENTIFIER}`,
  synonymsGeneric: `search_analyzer_synonyms_generic`,
  stemmedDelimiter: `search_analyzer_stemmed_delimiter_${MARKET_IDENTIFIER}`,
  delimiter: `search_analyzer_delimiter_${MARKET_IDENTIFIER}`,
  delimiterGeneric: `search_analyzer_delimiter_generic`,
} as const;

export type SearchTimeAnalyzer = ValueOf<typeof searchTimeAnalyzer>;

export type SearchableField<SubField extends ShopsSubFields> = {
  subField?: SubField;
  boost?: number;
  searchTimeAnalyzer?: SearchTimeAnalyzer;
};

export type SearchableFields<Field extends ShopsSearchableField> = {
  field: Field;
  boost?: number;
  phraseMatchBoost?: number; // to be specified if phrase match boosting is to be applied to this field
  nestedPath?: string; // to be specified if a nested query is to be performed on this field
  disMaxBlock?: string;
};

export const shopsSearchableField = {
  name: 'searchableText.name',
} as const;

export const shopsSubFields = {
  edgeNgram: 'edge_ngram',
  ngram: 'ngram',
  shingles: 'shingles',
} as const;

export type ShopsSearchableField = ValueOf<typeof shopsSearchableField>;
export type ShopsSubFields = ValueOf<typeof shopsSubFields>;

export type ShopsSearchField<PossibleEmpties = undefined> =
  SearchableFields<ShopsSearchableField> & {
    searchables: {
      exact?: SearchableField<ShopsSubFields> | PossibleEmpties;
      edgeNgram?:
        | SearchableField<(typeof shopsSubFields)['edgeNgram']>
        | PossibleEmpties;
      ngram?:
        | SearchableField<(typeof shopsSubFields)['ngram']>
        | PossibleEmpties;
      shingles?:
        | SearchableField<(typeof shopsSubFields)['shingles']>
        | PossibleEmpties;
    };
  };

```

## Possible Solution

Fix it here:
https://github.com/lukeautry/tsoa/blob/master/packages/cli/src/metadataGeneration/typeResolver.ts#L1104

This should never resolve to invalid name, should **always** output according to acceptable pattern `"^[a-zA-Z0-9\.\-_]+$"`

## Steps to Reproduce

1. 2. 3. 4.

## Context (Environment)

Version of the library: `5.1.1`
Version of NodeJS: `v18.16.0`
Yarn version: `1.22.19`

## Detailed Description

Basically we need to do type-gymnastics to resolve the types properly in TSOA.
This causes it to generate bad names which is not accepted by the .

**Comment by KennyLindahl** [comment]
This is still an issue, please remove the `stale` label

**Comment by NoodleOfDeath** [comment]
Can this please be fixed? I have to use a sed work around for now

**Comment by micahriggan** [comment]
I had an issue where some of my prisma types have . or - in the type names. After generating the spec and routes, I run this script to patch the swagger.json file 

```javascript
const fs = require('fs');
const path = require('path');

const swaggerFilePath = path.join(__dirname, '../build/swagger.json');

// Read the swagger.json file as a string
fs.readFile(swaggerFilePath, 'utf8', (err, data) => {
  if (err) {
    console.error('Error reading the file:', err);
    return;
  }

  // Parse the JSON
  const swagger = JSON.parse(data);

  // Build a list of schemas that need to be replaced
  const schemasToReplace = Object.keys(swagger.components.schemas)
    .filter(schema => schema.includes('.') || schema.includes('-'))
    .map(schema => ({
      original: schema,
      updated: schema.replace(/[.-]/g, '_')
    }));

  // Perform full-file text string replace for each schema
  let updatedData = data;
  schemasToReplace.forEach(({ original, updated }) => {
    const regex = new RegExp(`${original}`, 'g');
    updatedData = updatedData.replace(regex, `${updated}`);
  });

  // Write the updated data back to the file
  fs.writeFile(swaggerFilePath, updatedData, 'utf8', (err) => {
    if (err) {
      console.error('Error writing the file:', err);
      return;
    }
    console.log('swagger.json has been updated successfully. Replaced schemas:', schemasToReplace);
  });
});
```
